### ISSUE TITLE
document explicitly what a `ATTR{foo}!="bar"` check means for devices that have no attribute "foo"

### ISSUE THREAD

**Issue by Malvineous** [systemd/systemd#15984]
**systemd version the issue has been seen with**
> systemd 245 (245.2-2-arch)
> +PAM +AUDIT -SELINUX -IMA -APPARMOR +SMACK -SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +SECCOMP +BLKID +ELFUTILS +KMOD +IDN2 -IDN +PCRE2 default-hierarchy=hybrid

**Used distribution**
 > Arch Linux

**Expected behaviour you didn't see**
 > I expected to be able to connect to the user session DBus, which ultimately allows the LightDM display manager to present a login box.  Any attempts to connect to the session bus time out or never return, causing any program that relies on DBus (like LightDM) to freeze.

**Unexpected behaviour you saw**
 > The display manager did not show a login box, which I have traced to the user DBus session not responding.

Using `dbus-monitor` with `strace` shows the same problem as running LightDM through `strace`, in that the program tries to authenticate with DBus but never receives a response:
```
$ strace dbus-monitor --address unix:path=/run/user/620/bus
connect(3, {sa_family=AF_UNIX, sun_path="/run/user/620/bus"}, 19) = 0
sendto(3, "\0", 1, MSG_NOSIGNAL, NULL, 0) = 1
sendto(3, "AUTH EXTERNAL 30\r\n", 18, MSG_NOSIGNAL, NULL, 0) = 18
poll([{fd=3, events=POLLIN}], 1, -1
```
It just sits there forever waiting for that `poll()` to return.  If I use the system bus instead then it works fine, getting a response from the `AUTH EXTERNAL` message and continuing on:
```
$ strace dbus-monitor --address unix:path=/var/run/dbus/system_bus_socket
connect(3, {sa_family=AF_UNIX, sun_path="/var/run/dbus/system_bus_socket"}, 33) = 0
sendto(3, "\0", 1, MSG_NOSIGNAL, NULL, 0) = 1
sendto(3, "AUTH EXTERNAL 30\r\n", 18, MSG_NOSIGNAL, NULL, 0) = 18
poll([{fd=3, events=POLLIN}], 1, -1)    = 1 ([{fd=3, revents=POLLIN}])
read(3, "OK 682a5ae41ee048ed8f2f2b585ec13"..., 2048) = 37
poll([{fd=3, events=POLLOUT}], 1, -1)   = 1 ([{fd=3, revents=POLLOUT}])
sendto(3, "NEGOTIATE_UNIX_FD\r\n", 19, MSG_NOSIGNAL, NULL, 0) = 19
poll([{fd=3, events=POLLIN}], 1, -1)    = 1 ([{fd=3, revents=POLLIN}])
read(3, "AGREE_UNIX_FD\r\n", 2048)      = 15
```
The problem seems to be that the user session DBus is not responding.  I am not sure how to debug this further (see below).

**Steps to reproduce the problem**

    $ DBUS_SESSION_BUS_ADDRESS=unix:path=/run/user/620/bus systemctl --user status
    Failed to connect to bus: No such file or directory    # As expected
    $ systemctl start lightdm
    $ DBUS_SESSION_BUS_ADDRESS=unix:path=/run/user/620/bus systemctl --user status
    Failed to read server status: Connection timed out

**Other details**
```
$ busctl | grep lightdm
NAME                               PID PROCESS         USER            CONNECTION    UNIT                     SESSION DESCRIPTION
:1.467                         1332156 systemd         lightdm         :1.467        user@620.service         -       -          
:1.496                         1334386 lightdm         root            :1.496        lightdm.service          -       -          
:1.497                         1334391 Xorg            root            :1.497        lightdm.service          -       -          
:1.498                         1334399 lightdm         root            :1.498        session-c23.scope        c23     -          
org.freedesktop.DisplayManager 1334386 lightdm         root            :1.496        lightdm.service          -       -          

$ ps -h -o cmd -q 1332156
/usr/lib/systemd/systemd --user
```
Attaching to the `user@620.service` (`systemd --user`) PID with gdb shows:
```
(gdb) bt
#0  0x00007fe193ca0b5f in ppoll () from /usr/lib/libc.so.6
#1  0x00007fe1939f3291 in ?? () from /usr/lib/systemd/libsystemd-shared-245.so
#2  0x00007fe1939f56c9 in bus_ensure_running () from /usr/lib/systemd/libsystemd-shared-245.so
#3  0x00007fe1939f5763 in sd_bus_flush () from /usr/lib/systemd/libsystemd-shared-245.so
#4  0x000055926e375818 in ?? ()
#5  0x000055926e34b2bf in ?? ()
#6  0x000055926e3214dc in ?? ()
#7  0x000055926e33d900 in ?? ()
#8  0x000055926e341b97 in ?? ()
#9  0x000055926e34c697 in ?? ()
#10 0x00007fe1939c5442 in ?? () from /usr/lib/systemd/libsystemd-shared-245.so
#11 0x00007fe1939c588b in sd_event_dispatch () from /usr/lib/systemd/libsystemd-shared-245.so
#12 0x00007fe1939c7311 in sd_event_run () from /usr/lib/systemd/libsystemd-shared-245.so
#13 0x000055926e38701f in ?? ()
#14 0x000055926e2d9156 in ?? ()
#15 0x00007fe193bd3023 in __libc_start_main () from /usr/lib/libc.so.6
#16 0x000055926e2da49e in ?? ()
```
It seems to be always stuck in the [`ppoll()`](https://github.com/systemd/systemd/blob/cf3317f63a24c17ca06a5c277171550c28ed7ef2/src/libsystemd/sd-bus/sd-bus.c#L3121) call for some reason.  It would make sense if it was waiting for a message to arrive, but why is this wait inside a call to `sd_bus_flush()`?  That suggests it hasn't finished processing an earlier message?

If there is anything else I can do to diagnose this further please let me know, as I'm not sure where to look next.

**Comment by poettering** [comment]
is this with dbus-daemon or dbus-broker?

Does the dbus-daemon/dbus-broker --user service get started? Does it show up in ps?  Can you provide logs, please? Have you turned on debug logging for user@.service? (i.e. SYSTEMD_LOG_LEVEL=debug env var on user@.service?

**Comment by Malvineous** [comment]
**TL;DR: The DBus problem is ultimately caused by systemd-udevd creating device nodes in `/dev` with the wrong permissions and I can't work out why, as the incorrect nodes aren't listed in any `.rules` files.  If I fix them, they just get changed back to the wrong permissions each reboot.**

---

Many thanks for your suggestions!

After all my debugging it never even occurred to me to check the logs of the service itself, how silly of me.  I'm using dbus-daemon and the logs suggest it can't start:
```
Jun 03 22:12:57 korath2 systemd[1341246]: Started D-Bus User Message Bus.
Jun 03 22:12:57 korath2 systemd[1905976]: dbus.service: Failed to set up standard input: Permission denied
Jun 03 22:12:57 korath2 systemd[1905976]: dbus.service: Failed at step STDIN spawning /usr/bin/dbus-daemon>
Jun 03 22:12:57 korath2 systemd[1341246]: dbus.service: Main process exited, code=exited, status=208/STDIN
Jun 03 22:12:57 korath2 systemd[1341246]: dbus.service: Failed with result 'exit-code'.
Jun 03 22:14:27 korath2 systemd[1341246]: Started D-Bus User Message Bus.
Jun 03 22:14:27 korath2 systemd[1906119]: dbus.service: Failed to set up standard input: Permission denied
Jun 03 22:14:27 korath2 systemd[1906119]: dbus.service: Failed at step STDIN spawning /usr/bin/dbus-daemon>
Jun 03 22:14:27 korath2 systemd[1341246]: dbus.service: Main process exited, code=exited, status=208/STDIN
Jun 03 22:14:27 korath2 systemd[1341246]: dbus.service: Failed with result 'exit-code'.
```
I investigated `Failed to set up standard input` but I could only find posts advising to disable SELinux (I'm not running SELinux) or to check TTY permissions (tested with `chmod 777 /dev/tty*` but it made no difference).  So I ran `systemctl edit user@620.service` and set it to:
```
[Service]
Environment=SYSTEMD_LOG_LEVEL=debug
```
To enable debugging as suggested.  Restarting lightdm and checking the logs for `user@620.service` provided some more detail:
```
dbus.service: Failed to set 'memory.limit_in_bytes' attribute on '/user.slice/user-620.slice/user@620.service/dbus.service' to '-1': No such file or directory
Failed to read pids.max attribute of cgroup root, ignoring: No such file or directory
dbus.service: Failed to set 'pids.max' attribute on '/user.slice/user-620.slice/user@620.service/dbus.service' to '77190': No such file or directory
dbus.service: Passing 1 fds to service
dbus.service: About to execute: /usr/bin/dbus-daemon --session --address=systemd: --nofork --nopidfile --systemd-activation --syslog-only
dbus.service: Forked /usr/bin/dbus-daemon as 1909612
Skipping PR_SET_MM, as we don't have privileges.
dbus.service: Changed dead -> running
dbus.service: Job 19 dbus.service/start finished, result=done
Started D-Bus User Message Bus.
Bus bus-api-user: changing state UNSET ? OPENING
Bus bus-api-user: changing state OPENING ? AUTHENTICATING
dbus.service: Failed to set up standard input: Permission denied
dbus.service: Failed at step STDIN spawning /usr/bin/dbus-daemon: Permission denied
Successfully connected to API bus.
Received SIGCHLD from PID 1909612 ((s-daemon)).
Child 1909612 ((s-daemon)) died (code=exited, status=208/STDIN)
dbus.service: Failed to read oom_kill field of memory.events cgroup attribute: No such file or directory
dbus.service: Child 1909612 belongs to dbus.service.
dbus.service: Main process exited, code=exited, status=208/STDIN
dbus.service: Failed with result 'exit-code'.
dbus.service: Service will not restart (restart setting)
dbus.service: Changed running -> failed
dbus.service: Unit entered failed state.
```
I tried running the failing dbus command as the same user:
```
$ su lightdm -s /bin/bash
$ /usr/bin/dbus-daemon --session --address=systemd: --nofork --nopidfile --systemd-activation --syslog-only
dbus-daemon: fatal error setting up standard fds: Failed to open /dev/null: Permission denied
$ ls -ls /dev/null
crw-rw---- 1 root users 1, 3 May 17 22:56 /dev/null
```
Well apparently that's the problem.  Looking on another PC, the permissions are meant to be:
```
crw-rw-rw- 1 root root 1, 3 Jan 21 22:21 /dev/null
```
It's not just `/dev/null` either, `/dev/urandom` and a bunch of others have the wrong permissions - fixing one just makes `dbus-session` complain about another file in `/dev`.

So it looks like it's actually an issue with `systemd-udevd`.  I tried fixing the permissions like this:
```
$ udevadm trigger --sysname-match=null
$ ls /dev/null
crw-rw-rw- 1 root users 1, 3 Jun  3 22:42 /dev/null
```
Which fixed the mode but left the group wrong.  However when I reboot the machine, the permissions go back to the wrong ones that they originally were!

I confirmed `/dev` is a `devtmpfs` mount, so the files are not persisting over a reboot, systemd-udevd is actually recreating all the device nodes with the wrong permissions on each boot.

Some Googling suggests these rules are stored in `/usr/lib/udev/rules.d/50-udev-default.rules` however grep can't find any mention of `null` or `urandom` in any of those rules files - either on the broken system, or on the working system this one was imaged from.

Where are the rules stored for `/dev/null` and `/dev/urandom` and what would cause systemd-udevd to create these files with the wrong permissions?

**Comment by Malvineous** [comment]
Some help on the Arch Linux forum has revealed that the problem is caused by these rules in `/etc/udev/rules.d`:
```
# Make normal users able to capture events coming off the mouse
ENV{ID_VENDOR_ID}=="046d", ENV{ID_MODEL_ID}=="c525", GROUP="users", MODE="0660"
ENV{ID_VENDOR_ID}=="046d", ENV{ID_MODEL_ID}=="c52b", GROUP="users", MODE="0660"
```
Any idea why that rule would have worked for years but only recently decided to match everything from `/dev/null` to `/dev/tty*`?

**Comment by poettering** [comment]
this indeed looks like a bug in rule processing. @yuwata?

**Comment by yuwata** [comment]
@Malvineous I cannot reproduce the issue. Could you boot with `udev.log_priority=debug` and provides the relevant logs for `/dev/null` or `/dev/tty*`?
Also, please provide the results of `udevadm info /devices/virtual/mem/null` and `udevadm test /devices/virtual/mem/null`.

**Comment by poettering** [comment]
I tried hard to reproduce this too and couldn't. The rules snippets you provided didnt trigger any such issue here.

Is it possible that some other rule first assigned the two ENV{} props to all devices by accident?

**Comment by Malvineous** [comment]
Really sorry guys, you're right, in my enthusiasm I accidentally changed too many things at once and reported the wrong info.  Going back and having a much more methodical look at how to reproduce the problem, the actual rule at fault is this one:
```
ATTR{name}!="logid", GOTO="logid_end"
ACTION=="add", GROUP="users", MODE="0660", SYMLINK+="input/by-id/logid"
LABEL="logid_end"
```
Putting that as the only file in `/etc/udev/rules.d/` is enough to trigger the issue for me.  Presumably I've done something wrong with the condition for that `GOTO`.

I think I may have further confused myself because looking at the dates, the system I was using this on has a long enough uptime that I suspect I haven't actually rebooted since adding this rule in.  I tested that it created the intended symlink but of course didn't test if it changed `/dev/null` or anything else.  This means it is probably not a bug introduced with a recent version of systemd, but rather just me forgetting that I haven't tested the above rule through a reboot, at least until I went to migrate the install to new hardware.  Sorry about this.

I'm not quite sure what I've done wrong with the condition though - does it fail if `ATTR{name}` doesn't exist?

**Comment by poettering** [comment]
```
ATTR{name} != "logid"
```

That check doesn't mean "the device's attribute 'name' is not 'logid'" as one might expect but instead means "not(the device's attribute 'name' is 'logid')".

And ovbviously, the latter is true for /dev/null: it has no attribute 'name' which hence also isn't "logind"...

We could certainly make this clearer in the docs however

**Comment by poettering** [comment]
I converted this hence into a documenation bug

**Comment by boucman** [comment]
so, how would you test that an attribute exist ? is that what TEST= is for ?

**Comment by arvidjaar** [comment]
> so, how would you test that an attribute exist ? is that what TEST= is for ?

Yes, if `TEST` gets relative filename, it is assumed to be under device path in syfs (i.e. device attribute). Which incidentally is not documented either :) 

Alternatively `ATTR{foo} != ""` probably works as well unless empty attribute is valid.
